src/main/java/de/neemann/digital/core/io/Port.java

```diff
@@ -0,0 +1,219 @@
+     */
+    public Port(ElementAttributes attributes) {
+        portMode = attributes.get(Keys.PORT_MODE);
+        portTelnet = attributes.get(Keys.PORT_TELNET);
+        label = attributes.getLabel();
+
+        if (portMode) {
+            sout = new ObservableValue("out", 1).setPinDescription(DESCRIPTION);
+        } else {
+            pout = new ObservableValue("Q", 8).setPinDescription(DESCRIPTION);
+            bsy = new ObservableValue("bsy", 1).setPinDescription(DESCRIPTION);
+            avail = new ObservableValue("avail", 1).setPinDescription(DESCRIPTION);
+        }
+        if (portSocket == null) {
+            portSocket = new PortSocket(this, 2323, portTelnet);
+            portSocket.start();
+        }
+    }
+
+    @Override
+    public void setInputs(ObservableValues inputs) throws NodeException {
+        if (portMode) {
+            sin = inputs.get(0).addObserverToValue(this).checkBits(1, this, 0);
+            clock = inputs.get(1).addObserverToValue(this).checkBits(1, this, 1);
+        } else {
+            pin = inputs.get(0).addObserverToValue(this).checkBits(8, this, 0);
+            clock = inputs.get(1).addObserverToValue(this).checkBits(1, this, 1);
+            stb = inputs.get(2).addObserverToValue(this).checkBits(1, this, 2);
+            ack = inputs.get(3).addObserverToValue(this).checkBits(1, this, 3);
+        }
+    }
+
+    @Override
+    public ObservableValues getOutputs() {
+        if (portMode) {
+            return new ObservableValues(sout);
+        } else {
+            return new ObservableValues(pout, bsy, avail);
+        }
+    }
+
+
+    @Override
+    public void readInputs() throws NodeException {
+        boolean nowClock = clock.getBool();
+
+        if (!portMode) {
+            boolean nowStb = stb.getBool();
+            boolean nowAck = ack.getBool();
+
+            // If rising edge on stb then send the paralell-in (pin) to
+            // the socket, and set the bsy flag (it will be automatically
+            // be cleared in X clock cycles to simulate a slow external device)
+            if (lastStb != nowStb) {
+                if (nowStb) {
+                    if (!bsy.getBool()) {
+                        toSocket((int) pin.getValue());
+                        bsy.setValue(1);
+                        bsyCnt=100;
+                    }
+                }
+                lastStb=nowStb;
+            }
+
+            // Check if it's time to clear the bsy flag
+            if (!portMode) {
+                if (lastClock != nowClock) {
+                    if (bsyCnt>0) {
+                        if (--bsyCnt==0) bsy.setValue(0);
+                    }
+                }
+            }
+
+            // If there's pending data in the buffer and the last
+            // data has been ack'ed then update the Q port
+            if (!buf.isEmpty() && !avail.getBool()) {
+                poutVal=buf.remove();
+                pout.set((long) poutVal, 0);
```

Comment:
Never write to an output in the readInputs() method.

Reply:
That was esoteric knowledge ;-) But seems very logical.